pkg/imagestream/imagestream_test.go

```diff
@@ -0,0 +1,423 @@
+			client:              false,
+		},
+	}
+
+	for _, testCase := range testCases {
+		// Pre-populate the runtime objects
+		var runtimeObjects []runtime.Object
+
+		var testSettings *clients.Settings
+
+		testImageStream := generateImageStream(testCase.name, testCase.namespace)
+
+		if testCase.addToRuntimeObjects {
+			runtimeObjects = append(runtimeObjects, testImageStream)
+		}
+
+		if testCase.client {
+			testSettings = clients.GetTestClients(clients.TestClientParams{
+				K8sMockObjects:  runtimeObjects,
+				SchemeAttachers: testSchemes,
+			})
+		}
+
+		builderResult, err := Pull(testSettings, testCase.name, testCase.namespace)
+		assert.Equal(t, testCase.expectedError, err)
+
+		if testCase.expectedError == nil {
+			assert.Equal(t, testImageStream.Name, builderResult.Object.Name)
+			assert.Equal(t, testImageStream.Namespace, builderResult.Object.Namespace)
+			assert.Nil(t, err)
+		}
+	}
+}
+
+func TestImageStreamGet(t *testing.T) {
+	testCases := []struct {
+		testImageStream *Builder
+		expectedError   error
+	}{
+		{
+			testImageStream: buildValidImageStreamBuilder(buildImageStreamClientWithDummyObject()),
+			expectedError:   nil,
+		},
+		{
+			testImageStream: buildInValidImageStreamBuilder(buildImageStreamClientWithDummyObject()),
+			expectedError:   fmt.Errorf("the imageStream 'name' cannot be empty"),
+		},
+		{
+			testImageStream: buildValidImageStreamBuilder(clients.GetTestClients(clients.TestClientParams{})),
+			expectedError:   fmt.Errorf("imagestreams.image.openshift.io \"network-tools\" not found"),
+		},
+	}
+
+	for _, testCase := range testCases {
+		imageStreamObj, err := testCase.testImageStream.Get()
+
+		if testCase.expectedError == nil {
+			assert.Equal(t, imageStreamObj.Name, testCase.testImageStream.Definition.Name)
+			assert.Equal(t, imageStreamObj.Namespace, testCase.testImageStream.Definition.Namespace)
+			assert.Nil(t, err)
+		} else {
+			assert.Equal(t, testCase.expectedError.Error(), err.Error())
+		}
+	}
+}
+
+func TestImageStreamExists(t *testing.T) {
+	testCases := []struct {
+		testImageStream *Builder
+		expectedStatus  bool
+	}{
+		{
+			testImageStream: buildValidImageStreamBuilder(buildImageStreamClientWithDummyObject()),
+			expectedStatus:  true,
+		},
+		{
+			testImageStream: buildInValidImageStreamBuilder(buildImageStreamClientWithDummyObject()),
+			expectedStatus:  false,
+		},
+		{
+			testImageStream: buildValidImageStreamBuilder(clients.GetTestClients(clients.TestClientParams{})),
+			expectedStatus:  false,
+		},
+	}
+
+	for _, testCase := range testCases {
+		exist := testCase.testImageStream.Exists()
+		assert.Equal(t, testCase.expectedStatus, exist)
+	}
+}
+
+//nolint:funlen
+func TestImageStreamGetDockerImage(t *testing.T) {
+	testCases := []struct {
+		testImageStreamClientWithDummyObject *imagev1.ImageStream
+		testImageTag                         string
+		expectedError                        error
+	}{
+		{
+			testImageStreamClientWithDummyObject: &imagev1.ImageStream{
```

Comment:
since a lot of this gets repeated, could you please create a function like `buildDummyImageStreamWithTags(tags []imagev1.TagReference) *imagev1.ImageStream`?

Reply:
yes, sure, done